✨ Add Readout Format for State Vector and Probabilities

ystade opened the pull request:
### Issue

As discussed in #80, more readout formats are desired, esp., in the case when the QDMI device is a simulator.

### Solution

This PR adds the readout types and implements the corresponding behaviour in both example devices.
```C
QDMI_JOB_RESULT_STATEVECTOR_DENSE,
QDMI_JOB_RESULT_PROBABILITIES_DENSE,
QDMI_JOB_RESULT_STATEVECTOR_SPARSE_KEYS,
QDMI_JOB_RESULT_STATEVECTOR_SPARSE_VALUES,
QDMI_JOB_RESULT_PROBABILITIES_SPARSE_KEYS,
QDMI_JOB_RESULT_PROBABILITIES_SPARSE_VALUES,
```

ystade committed:
✨ Add new result formats

ystade committed:
✨ Support new result formats in C and C++ example devices

ystade committed:
✅ Add test for new readout types

ystade committed:
📝 Prettify docs

ystade committed:
🎨 Substitute i++ by ++i where possible for consistency

burgholzer committed:
✅ improved test structure and amount of testing

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer commented on examples/device/c/device.c:
```diff
@@ -426,92 +445,254 @@ int QDMI_control_get_data_dev(QDMI_Job job, const QDMI_Job_Result result,
+    }
+    return QDMI_SUCCESS;
+  }
+  if (result == QDMI_JOB_RESULT_STATEVECTOR_SPARSE_KEYS) {
+    int dense_size = 0;
+    QDMI_control_get_data_dev(job, QDMI_JOB_RESULT_STATEVECTOR_DENSE, 0, NULL,
+                              &dense_size);
+    double *dense_data = malloc(dense_size);
+    QDMI_control_get_data_dev(job, QDMI_JOB_RESULT_STATEVECTOR_DENSE,
+                              dense_size, dense_data, NULL);
+    int num_qubits = 0;
+    QDMI_query_device_property_dev(QDMI_DEVICE_PROPERTY_QUBITSNUM, sizeof(int),
+                                   &num_qubits, NULL);
+    // count non-zero elements
+    int count = 0;
+    for (size_t i = 0; i < 1ULL << num_qubits; i++) {
+      if (dense_data[2 * i] != 0.0 || dense_data[(2 * i) + 1] != 0.0) {
+        count++;
+      }
+    }
+    if (data != NULL) {
```
The sparse functions for the state vector and the probability vector still feature the kind of code duplication that was also present in the histogram output before this PR.
Would be great if we could also reduce the duplication here.
See https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/82/commits/3d396d51b3c901602fc3e137ff8cd42fca835145 for more details on what I changed up in the histogram results implementation.

Another important point in that regard: in most of these routines, the call to the `DENSE` variant can be omitted and the `job->state_vec` data can be directly used.

This comment holds for both the C and the C++ device.

ystade replied:
Is that desired to refer to `job->state_vec` and respective data structures rather than calling the `DENSE` function? Calling the dense function is more generic and also works if one decides to store the results somewhere or somehow different.

burgholzer replied:
I believe that it is fine to use whatever the job stores as it's results. This is up to the device implementer anyway. We have to keep in mind that not all the devices will be able to provide all of the results for a given job.
This will most likely require a way to query, which kind of results are supported by the device and then explicitly requesting them as part of the job (parameters).
But I guess this is something for a different PR.

I'd generally try to keep the example device implementations as simple as possible.

burgholzer replied:
Another argument: I wouldn't like to set the precedent that it is common practice to call the dense methods as part of the sparse methods. The sparse methods are there for a reason: in many practical cases you won't have the memory in all the world to store the dense result, but you might be able to directly report back the sparse result.
Us even being able to implement it in this fashion is simply due to the fact that we are only looking at a small example use case and that we generate a random result in the form of a random dense state-vector.

I would expect that practical implementations of simulators actually would implement this the other way round in many cases. That is, they offer the sparse result and, optionally, a way to create a dense result from that assuming that enough memory is available.

ystade replied:
True, agreed.

burgholzer replied:
https://github.com/Munich-Quantum-Software-Stack/QDMI/pull/82/commits/ab65825961ba8e5b7932d96b71692151754550b2 now contains these refactors and considerably shrinks the implementation. I think it looks quite nice now.
Given all tests pass, I'd be up for merging this.

burgholzer reviewed (changes requested):
Thanks for your work on this 🙏🏼 
I added a bit of testing functionality and fixed a couple of bugs in the implementation.
I also rebased this on top of #84.

As a consequence, clang-tidy does now complain quite a bit more than before 🙃 (mainly cause by the `qdmi::warnings` target now being linked to the devices. However, all of these should be easily addressed.

Finally, the parts that codecov claims are uncovered, are mostly the "failure" branches, where the provided buffers are too small.
It would be great, if we could also add tests for these cases. Especially, to demonstrate the expected behavior.

Given the amount of conversion warnings here, I am thinking whether it would maybe be better to get a fix for #81 in first and rebase on top of that. 
@ystade what do you think? 
I can work on #81 tomorrow.

ystade commented:
> Thanks for your work on this 🙏🏼 I added a bit of testing functionality and fixed a couple of bugs in the implementation. I also rebased this on top of #84.
> 
> As a consequence, clang-tidy does now complain quite a bit more than before 🙃 (mainly cause by the `qdmi::warnings` target now being linked to the devices. However, all of these should be easily addressed.
> 
> Finally, the parts that codecov claims are uncovered, are mostly the "failure" branches, where the provided buffers are too small. It would be great, if we could also add tests for these cases. Especially, to demonstrate the expected behavior.
> 
> Given the amount of conversion warnings here, I am thinking whether it would maybe be better to get a fix for #81 in first and rebase on top of that. @ystade what do you think? I can work on #81 tomorrow.

Yes, I am in favour of first addressing #81 and rebasing this. This saves a couple of static_casts or warnings.

burgholzer commented:
Alright, I rebased on top of #85 (which addressed #81) and cleaned up the commit history here.
Modulo some code quality improvements, this should be ready to go 🚀

burgholzer committed:
✅ add tests for buffers that are too small

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 reduce code duplication and improve implementation

Signed-off-by: burgholzer <burgholzer@me.com>

burgholzer committed:
🎨 more compact code

Signed-off-by: burgholzer <burgholzer@me.com>

ystade commented:
I skimmed very quickly over the changes and it looks good to me. If you do not want a rigorous review from me and the quick skimming suffices, it is also a go from my side.

burgholzer commented:
> I skimmed very quickly over the changes and it looks good to me. If you do not want a rigorous review from me and the quick skimming suffices, it is also a go from my side.

That'a sufficient. Many thanks! 
Everything else can be improved in a follow up 😌
